### PR TITLE
feat: add better handling when play key fails to fetch

### PR DESCRIPTION
**Pull request by JLaferri**
Previously when the user had no internet access or Slippi servers were down we would show the online activation step which was pretty confusing to users.

New behavior:
https://user-images.githubusercontent.com/1534726/153937946-e748898c-87fd-4f17-a0fc-5bd83fbbaacb.mp4

### CODE REVIEW

File: src/renderer/containers/Header/UserInfo.tsx [hunk 1]
```diff
@@ -11,8 +11,19 @@ export const UserInfo: React.FC<{
   uid: string;
   displayName: string | null;
   playKey: PlayKey | null;
+  isServerError: boolean | null;
   loading: boolean;
-}> = ({ uid, displayName, playKey, loading }) => {
+}> = ({ uid, displayName, playKey, isServerError, loading }) => {
+  const isErrorText = isServerError || !playKey;
```
Comment (NikhilNarayana) [inline src/renderer/containers/Header/UserInfo.tsx:17]:
```suggestion
  const showErrorText = serverError || !playKey;
```

File: src/renderer/lib/hooks/useAccount.ts [hunk 2]
```diff
@@ -34,10 +36,10 @@ export const useAccount = create(
 
         set({ loading: true });
         await fetchPlayKey()
-          .then((playKey) => set({ playKey }))
+          .then((playKey) => set({ playKey, isServerError: false }))
           .catch((err) => {
             console.warn("Error fetching play key: ", err);
-            set({ playKey: null });
+            set({ playKey: null, isServerError: true });
```
Comment (NikhilNarayana) [inline src/renderer/lib/hooks/useAccount.ts:42]: Ok so we actually need to think of another way to do this. A user with a brand new account triggers this bit of code and thus they are never able to set their connect code.

**Review by NikhilNarayana — CHANGES_REQUESTED** [review]
drop the `is` prefix on variables, nothing else in the launcher does it that way

File: src/renderer/lib/hooks/useAccount.ts (hunk 2 above)
Comment (JLaferri) [reply src/renderer/lib/hooks/useAccount.ts:42]: Huh interesting. In all likelihood I feel like an account without a play key should return a successful result with an empty play key, not an error. I can take a look tomorrow too.

Comment (JLaferri) [reply src/renderer/lib/hooks/useAccount.ts]: Addressed by https://github.com/project-slippi/slippi-launcher/pull/275/commits/ac47b418e12d4fc6cbacb68d6982f2ed3dfae9f4

The `connectCode` and `private` objects on `getUser` were null for a new user.

File: src/renderer/lib/slippiBackend.ts [hunk 3]
```diff
@@ -100,12 +100,20 @@ export async function fetchPlayKey(): Promise<PlayKey> {
...
   return {
     uid: user.uid,
-    connectCode: res.data.getUser.connectCode.code,
-    playKey: res.data.getUser.private.playKey,
-    displayName: res.data.getUser.displayName,
-    latestVersion: res.data.getLatestDolphin.version,
+    connectCode: connectCode,
+    playKey: playKey,
+    displayName: res.data.getUser?.displayName,
+    latestVersion: res.data.getLatestDolphin?.version,
```
Comment (vinceau) [inline src/renderer/lib/slippiBackend.ts:116]:
```suggestion
    connectCode,
    playKey,
    displayName,
    latestVersion,
```

This repo uses object shorthand notation so we should stick with that.

**Review by vinceau — CHANGES_REQUESTED** [review]

File: src/renderer/lib/slippiBackend.ts [hunk 4]
```diff
@@ -100,12 +100,20 @@ export async function fetchPlayKey(): Promise<PlayKey> {
 
   handleErrors(res.errors);
 
+  const connectCode = res.data.getUser?.connectCode?.code;
+  const playKey = res.data.getUser?.private?.playKey;
+  if (!connectCode || !playKey) {
```
[the remark this answers is not in the document]
Comment (JLaferri) [reply src/renderer/lib/slippiBackend.ts:105]: I agree with this from a type perspective, but not a functionality perspective. If we fail to get the version, we shouldn't be telling the user that they need to initialize their connect code.

I'd rather add null as a possible type for latest version and handle that correctly in the code.

[the remark this answers is not in the document]
Comment (JLaferri) [reply src/renderer/lib/slippiBackend.ts:105]: If an optional chain like that fails, does the value become null or undefined? Do ts types differentiate between the two?

File: src/renderer/lib/slippiBackend.ts (hunk 3 above)
Comment (JLaferri) [reply src/renderer/lib/slippiBackend.ts]: Addressed by https://github.com/project-slippi/slippi-launcher/pull/275/commits/fdabb249215b917821f9657d80dd191b3831b81d

File: src/renderer/lib/slippiBackend.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (JLaferri) [reply src/renderer/lib/slippiBackend.ts:106]: Addressed by https://github.com/project-slippi/slippi-launcher/pull/275/commits/fdabb249215b917821f9657d80dd191b3831b81d

**Review by vinceau — APPROVED** [review]
Not a big fan of how type-unsafe the GQL endpoints are but maybe we can fix that later. Also we should probably eventually fix the prop drilling that's happening in the `UserInfo` component but won't block the PR for it.

**Comment by JLaferri** [comment]
> Not a big fan of how type-unsafe the GQL endpoints are but maybe we can fix that later. Also we should probably eventually fix the prop drilling that's happening in the `UserInfo` component but won't block the PR for it.

GQL is never type unsafe by definition.

File: src/renderer/lib/hooks/useAccount.ts [hunk 5]
```diff
@@ -11,6 +11,7 @@ export const useAccount = create(
       user: null as firebase.User | null,
       loading: false,
       playKey: null as PlayKey | null,
+      serverError: false as boolean | null,
```
Comment (NikhilNarayana) [inline src/renderer/lib/hooks/useAccount.ts:14]: I don't see serverError ever being set to null or being set ambiguously, can we remove that extra type?

**Review by NikhilNarayana — CHANGES_REQUESTED** [review]

File: src/renderer/lib/hooks/useAccount.ts (hunk 5 above)
Comment (JLaferri) [reply src/renderer/lib/hooks/useAccount.ts:14]: Probably

**Review by NikhilNarayana — APPROVED** [review]
LGTM

**Pull request merged**